csp_billing_adapter_microsoft/plugin.py

```diff
@@ -142,18 +191,86 @@ def _fetch_metadata(url):
-   """
-    return json.loads(_fetch_metadata(TOKEN_URL))
+        log.error('Failed to retrieve metadata for: %s: %s', url, str(error))
+        return "{}"
+
+
+def _get_msi_token():
+    """Get the MSI token to authenticate when using the Billing API"""
+    # https://learn.microsoft.com/en-us/partner-center/marketplace/marketplace-metering-service-authentication
+
+    # Set resource id to the required value needed to to retrieve an
+    # MSI Authentication Token
+    resource = '20e940b3-4c77-4b0b-9a53-9e16a1b010a7'
+    client_id = os.environ['CLIENT_ID']
+    url = (
+        f"{METADATA_URL}"
+        f"identity/oauth2/token?api-version=2018-02-01&client_id={client_id}"
+        f"&resource={resource}"
+    )
+    try:
+        auth_token = json.loads(_fetch_metadata(url))
+        if auth_token["token_type"] == "Bearer" and auth_token["access_token"]:
+            return f'Bearer {auth_token["access_token"]}'
+        log.error('Invalid MSI token retrieved: %s', auth_token)
+        raise cba_exceptions.CSPBillingAdapterException
+    except ValueError as error:
+        log.error('Unable to acquire an MSI token %s:', str(error))
+        raise cba_exceptions.CSPBillingAdapterException from error
+
+
+def _create_usage_list(dimensions: dict, timestamp: datetime):
+    """Create the usage list used with the batchEventUsage API"""
+
+    usage = []
+    for dimension_name, quantity in dimensions.items():
+        if quantity == 0:
+            log.info(
+                'A "0" value was reported for %s, skipping meter billing',
+                dimension_name
+            )
+            continue
+
+        # Setup request body for the usage event API
+        usage.append(
+            {
+                'resourceUri': os.environ['EXTENSION_RESOURCE_ID'],
+                'quantity': quantity,
+                'dimension': dimension_name,
+                'effectiveStartTime': str(timestamp.astimezone(timezone.utc)),
+                'planId': os.environ['PLAN_ID']
+            }
+        )
+    return usage
+
+
+def _create_status_dict(response: dict):
+    """Create the status dict from the response from the batchUsageEvent API"""
+    status = {}
+    for resp in response["result"]:
+        if resp["status"] == "Accepted":
+            dim_status = {
+                "record_id": resp.get("usageEventId", None),
+                "status": "submitted"
+            }
+            log.info(
+                'New metered billing record added with ID %s:',
+                dim_status["record_id"]
+            )
+        else:
+            log.error(
+                'Unable to log metered billing record: %s',
+                resp
+            )
+            dim_status = {
+                "status": "failed",
+                "error":
+                    f'Failed to meter bill dimensions: '
+                    f'Status: {resp["status"]} '
+                    f'Message: {resp["error"]["message"]}'
```

Comment:
Are these keys certain to be in response? If not maybe worth using get with defaults.

Reply:
I believe these should be in the response based on the docs. I can add a failsafe to precent a KeyError

Reply:
@smarlowucf  please check this now.